prow/config/config.go

```diff
@@ -276,13 +276,40 @@ type Sinker struct {
 	MaxPodAge *metav1.Duration `json:"max_pod_age,omitempty"`
 }
 
+// LensConfig names a specific lens, and optionally provides some configuration for it.
+type LensConfig struct {
```

Comment:
One approach that is seen in k8s objects when we have many options for something that are distinct is an object with many fields, validated to have only one non-nil. Using that approach, we could have config like:

```yaml
lenses:
- metadata:
    required_files:
    - started.json
    optional_files:
    - finished.json
- buildlog:
    required_files:
    - build-log.txt
    highlight_regexes:
    - timed out
    - 'ERROR:'
    - (\s|^)(FAIL|Failure \[)\b
    - (\s|^)panic\b
    - ^E\d{4} \d\d:\d\d:\d\d\.\d\d\d]
- junit:
    required_files:
    - artifacts/junit.*\.xml
```

The Go structs would look like:

```go
lenses:
- metadata:
    required_files:
    - started.json
    optional_files:
    - finished.json
- buildlog:
    required_files:
    - build-log.txt
    highlight_regexes:
    - timed out
    - 'ERROR:'
    - (\s|^)(FAIL|Failure \[)\b
    - (\s|^)panic\b
    - ^E\d{4} \d\d:\d\d:\d\d\.\d\d\d]
- junit:
    required_files:
    - artifacts/junit.*\.xml

// LensConfig configures a lens. Only one component configuration may be set.
type LensConfig struct {
  Metadata MetadataLensConfig `json:"metadata,omitempty"`
  BuildLog BuildLogLensConfig `json:"buidlog,omitempty"`
  JUnit JUnitLensConfig `json:"junit,omitempty"`
}

// MetadataLensConfig configures the metadata lens
type MetadataLensConfig struct {
  LensFileConfig
}

func (c MetadataLensConfig) LensName() string {
  return "metadata"
}

// LensFileConfig is a single entry under Lenses, describing how to configure a lens
// to read a given set of files.
type LensFileConfig struct {
  // RequiredFiles is a list of regexes of file paths that must all be present for a lens to appear.
  // The list entries are ANDed together, i.e. all of them are required. You can achieve an OR
  // by using a pipe in a regex.
  RequiredFiles []string `json:"required_files"`
  // OptionalFiles is a list of regexes of file paths that will be provided to the lens if they are
  // present, but will not preclude the lens being rendered by their absence.
  // The list entries are ORed together, so if only one of them is present it will be provided to
  // the lens even if the others are not.
  OptionalFiles []string `json:"optional_files,omitempty"`
}
```

IMO the config turns out a little less verbose and more clear. What do you think?

Reply:
I agree that that's less verbose, and probably more legible. That said, I'm not a fan of the prow config parser knowing anything about what lenses exist (which is a requirement for that particular parsing setup) - ideally, they would be totally self-contained.

Reply:
>  ideally, they would be totally self-contained.

Why? We're going to need to parse them out anyway once config is loaded, and letting it happen in one pass would help things like `checkconfig` work as well

Reply:
It should be possible to make a spyglass lens without needing to touch anything outside the lens at all - this is the fundamental driver of most of spyglass's design decisions.

Forcing individual lens config to be mixed in with the rest of prow's config parsing undermines this design.

Reply:
Why is that a valuable goal, though? What does someone gain by editing `deck` source but not config?

Reply:
Ideally, one wouldn't be editing the deck source either - each lens is supposed to be self-contained and self-registering.

In practice, you _do_ have to touch deck source to force it to be linked in, but that is it. I don't want to expand that violation.

Eventually, we want to have remote lenses, which won't require touching _anything_ in core prow at all. It would be helpful if they could still be set up alongside the rest of them, which your proposal would preclude, but which is manageable with the setup as proposed.

Reply:
> Ideally, one wouldn't be editing the deck source either 

Don't we have to side-effect-import lenses right now?

But even if we have remote lenses, why do we want to shirk the type system and use `interface{}`, lose our ability to sanity check the configuration we have on disk and make the config more verbose? 

Do we not trust the `deck` deployment to do the work? Why would we prefer a remote lens that we send data and receive an IFrame from over in-process computation? Two extra network hops seems like a high cost for render time

Reply:
> Don't we have to side-effect-import lenses right now?

Yes, as I mentioned. This is regrettable, and I don't want to increase the encapsulation violation further.

> But even if we have remote lenses, why do we want to shirk the type system and use interface{}, lose our ability to sanity check the configuration we have on disk and make the config more verbose?

Because remote lenses do not require changing any source code in deck, prow, or this repo. Therefore they cannot, by definition, require that you make changes to prow's config parser.

> Do we not trust the deck deployment to do the work?

Remote lenses have always been part of Spyglass's design - not because we don't trust Deck (although give how much work some lenses are, perhaps having them be their own thing would be nice), but because we don't want them to be part of deck, or for working on them to require modifying deck, or for custom lenses to require a prow fork. This is a fundamental part of the Spyglass design - albeit one that has not yet reached fruition.

> Why would we prefer a remote lens that we send data and receive an IFrame from over in-process computation? Two extra network hops seems like a high cost for render time

Deck doesn't actually have most of the information that spyglass lenses use to render a PR, and where it does we tend to reacquire it by network requests to itself anyway. For what it's worth, my tentative in-my-head sketch for remote lenses shouldn't substantially increase the amount of network traffic (basically the lens gets some artifact metadata alongside a URL (e.g. signed GCS URL) that the lens can fetch to acquire the asset in its entirety)

Reply:
> Because remote lenses do not require changing any source code in deck, prow, or this repo. 

I guess I'm not super convinced this is an awesome end goal -- our attempts with things like external plugins are just clumsy and end up requiring an entire "microservice" deployment that takes an operation that needs milliseconds to complete and turns it into a network-hopping resource guzzler that needs to be deployed and monitored. 

If we have needs to support lenses that _cannot_ be part of `deck` then all this makes perfect sense, or if you think there will be special-purpose lenses that nobody else needs to run and the process of using external lenses won't be clunky.

Reply:
The original intent is that end-users of prow can make their own special lenses that they don't need or want to upstream, and they don't need to maintain a prow fork to do so. I don't think there's much that _cannot_ live in deck, but I question whether everything _should_ live in it (e.g. perhaps the buildlog lens, which IIRC is wildly expensive).

For a more real example, I know some people are/were working on a lens that renders output like https://apisnoop.cncf.io/, which would be a bizarrely specific thing to have in prow proper.

re: clunkiness: while an external lens would still need to be deployed somewhere, I don't think it should add significant request time overhead, and I think the development experience should not be dissimilar to the current one - I would envision being able to pull in a library that gives you the ~same API you get today.

Reply:
Sounds reasonable then